agent/secretsmgr/vault.go

```diff
@@ -0,0 +1,192 @@
+	"strings"
+	"time"
+
+	vault "github.com/hashicorp/vault/api"
+	"go.uber.org/zap"
+
+	"github.com/netboxlabs/orb-agent/agent/config"
+)
+
+var _ Manager = (*vaultManager)(nil)
+
+type vaultManager struct {
+	logger   *zap.Logger
+	config   config.VaultManager
+	ctx      context.Context
+	client   *vault.Client
+	usedVars map[string]cachedSecret
+	callback func([]string)
+	auth     authMethod
+	token    *vault.Secret
+}
+
+type cachedSecret struct {
+	Value     string         // The actual secret value
+	policyIDs map[string]any // The IDs of policies that have used this secret
+}
+
+func (v *vaultManager) Start(ctx context.Context) error {
+	v.ctx = ctx
+	v.usedVars = make(map[string]cachedSecret)
+
+	config := vault.DefaultConfig()
+
+	config.Address = v.config.Address
+	if v.config.Timeout == nil || *v.config.Timeout == 0 {
+		config.Timeout = 60 * time.Second
+	} else {
+		config.Timeout = time.Duration(*v.config.Timeout) * time.Second
+	}
+
+	if v.config.Auth == "" {
+		return fmt.Errorf("no auth method specified")
+	}
+
+	var err error
+	v.client, err = vault.NewClient(config)
+	if err != nil {
+		return err
+	}
+
+	if v.config.Namespace != "" {
+		v.client.SetNamespace(v.config.Namespace)
+	}
+
+	v.auth, err = newAuthentication(v.config.Auth, v.config.AuthArgs)
+	if err != nil {
+		return err
+	}
+
+	v.token, err = v.auth.vaultAuthenticate(ctx, v.client)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// RegisterUpdateCallback registers a callback function to be called when secrets are updated
+func (v *vaultManager) RegisterUpdateCallback(callback func([]string)) {
+	v.callback = callback
+}
+
+// SolveSecrets processes a policy payload and replaces vault references with environment variables
+func (v *vaultManager) SolveSecrets(payload config.PolicyPayload) (config.PolicyPayload, error) {
+	// Create a copy of the payload
+	newPayload := payload
+
+	// Process the Data field
+	processedData, err := v.processValue(payload.Data, payload.ID)
+	if err != nil {
+		return payload, err
+	}
+
+	newPayload.Data = processedData
+	return newPayload, nil
+}
+
+func (v *vaultManager) processValue(value any, id string) (any, error) {
+	switch val := value.(type) {
+	case string:
+		return v.processString(val, id)
+	case map[string]any:
+		return v.processMap(val, id)
+	case []any:
+		return v.processSlice(val, id)
+	default:
+		return val, nil
+	}
+}
+
+// processString processes a string and replaces vault references
+func (v *vaultManager) processString(s string, id string) (string, error) {
+	re := regexp.MustCompile(`\${vault://([^}]+)}`)
+	if !re.MatchString(s) {
+		return s, nil
+	}
+
+	match := re.FindStringSubmatchIndex(s)
+	if match == nil || len(match) < 4 {
+		return "", fmt.Errorf("failed to find vault reference in string: %s", s)
+	}
+
+	vaultPath := s[match[2]:match[3]]
+
+	if secrets, exists := v.usedVars[vaultPath]; exists {
+		secrets.policyIDs[id] = true
+		v.usedVars[vaultPath] = secrets
+		return secrets.Value, nil
+	}
```

Comment:
These are designed to be cached for the whole run of the process for now? Seems like we would ultimately like to be able to take advantage of the dynamic-ness of a backend like vault (just not sure how we've phased this in the roadmap)

Reply:
this is not a blocker, just a question, so I'm just going to resolve the convo here.